### PR TITLE
Enable group workspaces

### PR DESCRIPTION
**Pull request by CptTeddy**
Enable group workspaces

### CODE REVIEW

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/domain/api/review/ReviewApi.java [hunk 1]
```diff
@@ -53,12 +54,12 @@
...
      *
      * @param projectId   project id
-     * @param workspaceId workspace id
+     * @param workspace   workspace
      * @param title       review title
      * @param description review description
      * @return new review
      */
-    Review createReview(String projectId, String workspaceId, String title, String description);
+    Review createReview(String projectId, Workspace workspace, String title, String description);
```
Comment (kevin-m-knight-gs) [inline legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/domain/api/review/ReviewApi.java:62]: We should not use Workspace here. We should take the workspace id as a String and some other piece of information to indicate the workspace type. I think we should add a WorkspaceType enum in the model.

**Review by kevin-m-knight-gs — CHANGES_REQUESTED** [review]
This is an incomplete review, but I wanted to get these comments out.

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/application/review/CreateReviewCommand.java [hunk 2]
```diff
@@ -14,20 +14,22 @@
...
+
 public class CreateReviewCommand
 {
-    private String workspaceId;
+    private Workspace workspace;
     private String title;
     private String description;
 
-    public String getWorkspaceId()
+    public Workspace getWorkspace()
```
[the remark this answers is not in the document]
Comment (CptTeddy) [reply legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/application/review/CreateReviewCommand.java]: Agree -- have created WorkspaceType enum which is orthogonal to the WorkspaceAccessType enum and refactored along all lines of logic.

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/domain/api/review/ReviewApi.java (hunk 1 above)
Comment (CptTeddy) [reply legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/domain/api/review/ReviewApi.java]: Fixed as above mentioned!

**Comment by CptTeddy** [comment]
Note in addition to the above comments: have refactored default interface methods.

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectFileAccessProvider.java [hunk 3]
```diff
@@ -442,17 +451,36 @@ default FileModificationContext getWorkspaceFileModificationContext(String proje
...
+        WorkspaceType(final String label)
+        {
+            this.label = label;
+        }
+
+        public String getLabel()
+        {
+            return this.label;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (CptTeddy) [reply legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectFileAccessProvider.java:482]: Fixed: refactored as a public enum in model.

File: legend-sdlc-model/src/main/java/org/finos/legend/sdlc/domain/model/review/Review.java [hunk 4]
```diff
@@ -14,6 +14,7 @@
 
 package org.finos.legend.sdlc.domain.model.review;
 
+import org.finos.legend.sdlc.domain.model.project.workspace.Workspace;
```
[the remark this answers is not in the document]
Comment (CptTeddy) [reply legend-sdlc-model/src/main/java/org/finos/legend/sdlc/domain/model/review/Review.java:17]: Fixed: added WorkspaceType in Review model.

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/BaseGitLabApi.java [hunk 5]
```diff
@@ -173,38 +189,136 @@ protected static boolean isValidPackageableElementPath(String string)
...
+//            case GROUP_BACKUP:
+//            {
+//                throw new RuntimeException("Only non-group workspace access types should need to be adjusted, got: " + workspaceAccessType + " access type instead.");
+//            }
+//            default:
+//            {
+//                throw new RuntimeException("Unknown workspace access type " + workspaceAccessType);
+//            }
+//        }
+//    }
```
[the remark this answers is not in the document]
Comment (CptTeddy) [reply legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/BaseGitLabApi.java:312]: Apologies! Thanks and fixed.

File: legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabWorkspaceApi.java [hunk 6]
```diff
@@ -68,24 +69,28 @@ public GitLabWorkspaceApi(GitLabUserContext userContext, RevisionApi revisionApi
...
     @Override
-    public List<Workspace> getWorkspaces(String projectId)
+    public List<Workspace> getWorkspaces(String projectId, Set<ProjectFileAccessProvider.WorkspaceType> workspaceTypes)
     {
-        return this.getWorkspacesByAccessType(projectId, ProjectFileAccessProvider.WorkspaceAccessType.WORKSPACE);
+        return workspaceTypes.stream().map(type -> this.getWorkspacesByAccessType(projectId, type, ProjectFileAccessProvider.WorkspaceAccessType.WORKSPACE)).reduce(Collections.emptyList(), (a, b) ->
+        {
+            b.addAll(a);
+            return b;
+        });
```
[the remark this answers is not in the document]
Comment (CptTeddy) [reply legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabWorkspaceApi.java:78]: Fixed!

**Review by kevin-m-knight-gs — APPROVED** [review]

**Pull request merged**